Split VF2 `call_limit` to before and after first layout

jakelishman opened the pull request:
With the new VF2 scoring algorithm, the `max_trials` option is no longer appropriate as a limit, but is one that we functionally relied on for good runtimes in the face of high numbers of isomorphisms.  This splits the `call_limit` functionality in two, so that a separate (likely lower) limit can be placed on the time spent improving the layout, while retaining a high call limit for the initial search.

jakelishman committed:
Split VF2 `call_limit` to before and after first layout

With the new VF2 scoring algorithm, the `max_trials` option is no longer
appropriate as a limit, but is one that we functionally relied on for
good runtimes in the face of high numbers of isomorphisms.  This splits
the `call_limit` functionality in two, so that a separate (likely lower)
limit can be placed on the time spent improving the layout, while
retaining a high call limit for the initial search.

alexanderivrii commented:
Thanks @jakelishman. A few quick comments and questions while you are fixing conflicts.

The idea of specifying different resources for finding some solution and for improving on that solution makes a lot of sense. I don't feels strongly about this, but maybe instead of putting a total bound on the improvement-loop, we could put a call limit on each improvement-iteration, stopping the loop when no significant progress is made? (In other words, there is a potential difference between "we tried to improve an existing solution within `c` calls and failed" and "we keep finding significantly better and better solutions but had to stop because we reached `c` calls"). Also (and again I don't feel strongly about this), would it make sense to specify the two time-limits using different arguments rather than using a single tuple argument, what do you think?

And to confirm: after conflicts merge, `score_initial_layout` will not be a part of this PR, correct?

jakelishman commented:
> I don't feels strongly about this, but maybe instead of putting a total bound on the improvement-loop, we could put a call limit on each improvement-iteration, stopping the loop when no significant progress is made?

We could do this, but holistically, `call_limit` is supposed to be a deterministic variant on limiting the total runtime of the algorithm.  If the limit on the "improve" step is per-improvement, then it doesn't function as a total-runtime limiter any more.

> would it make sense to specify the two time-limits using different arguments rather than using a single tuple argument, what do you think?

I had it as the 2-tuple so that the existing API still continues to function in the same way as before (`call_limit=x` implies `call_limit=(x, x)`) since the normal "unset" value `None` has a special meaning in the argument (unlimited).  It's possible to do it another way, but since the two arguments are linked to each other, it felt marginally cleaner to me to expose them like that in the API.

alexanderivrii commented on crates/cext/src/transpiler/passes/vf2.rs:
```diff
@@ -162,10 +162,14 @@ pub unsafe extern "C" fn qk_vf2_layout_configuration_free(config: *mut VF2Layout
 ///
 /// @param config The configuration to update.
-/// @param limit The number of attempts to allow.  Set to a negative number to have no bound.
+/// @param before The number of attempts to allow before the first match is found.  Set to a
+///     negative number to have no bound.
+/// @param after The number of attempts to allow after the first match (if any) is found.  Set to a
+///     negative number to have no bound.
```
What is the expected behavior when the number of calls is `0`? If it's supposed to be `None` (as what I believe the `lift` closure below returns), then maybe change the above comment "set to a negative number" to "set to 0 or a negative number".

jakelishman replied:
Whoops, this just looks like a bug in `lift`.  Dunno what I was thinking.  Fixed in 8eb4927575.

alexanderivrii replied:
> /// @param after The number of attempts to allow after the first match (if any) is found.  Set to a
/// negative number to have no bound.

To me the above says something a bit different than 

> That could cause the limit to double in the case of call_limit = (1000, 1000) - you could use 999 trials, find a solution, then yours would cause you to gain another 1000 trials, but the intention is only to have 1.

like we do get this many additional trials. Maybe you can include this example in the docstring?

jakelishman replied:
I didn't include an exact example, but I drastically simplified the wording in 8e8fdce51b.

alexanderivrii commented on crates/transpiler/src/transpiler.rs:
```diff
@@ -136,16 +136,22 @@ pub fn transpile(
-        OptimizationLevel::Level1 | OptimizationLevel::Level2 => vf2::Vf2PassConfiguration {
-            call_limit: Some(5_000_000),
+        OptimizationLevel::Level1 => vf2::Vf2PassConfiguration {
+            call_limit: (Some(50_000), Some(1_000)),
```
This decreases the `call_limit` for optimization_level=1, is this intended? 

Also, how did you choose the default values for the improvement-loop budget (e.g., 100_000 for level 3)?

jakelishman replied:
For now I've reverted this in 72d0b158eab0f5cbd7dcaa354b8867776656b44a, but I assume I did it to cause it to match the Python-space numbers of `transpile` in the default pipeline.  We can/should do that in a separate PR, though.

I think I just made up the improvement values completely, tbh.  If you've got better suggestions for improving them, I'm open to it.

alexanderivrii replied:
Oh, I am completely fine with the numbers, I was just wondering if these were based looking at some particular benchmarks.

alexanderivrii commented on qiskit/transpiler/passes/layout/vf2_layout.py:
```diff
@@ -83,15 +83,26 @@ def __init__(
+                that the inner VF2 isomorphism will attempt to extend the mapping.  If ``None``,
+                then no limit.  If a 2-tuple, then the first value is the number of extensions
+                allowed when trying to find the first isomorphism, and the second is the budget for
+                extensions trying to improve the score.  Extensions made to find the first
```
The "budget for extensions" is the same thing as the "number of extensions" above, right?

Also, "Extensions made to find the first isomorphism also count against the budget for improvement" means that the extensions to find the first isomorphism do not count for the improvement-loop, right?

Update: actually I am now confused whether the first call limit only represents the budget to find the first layout, or whether it represents the total budget for the pass (including both finding the first layout and consecutively minimizing that).

jakelishman replied:
Iirc, my intention was that `call_limit.0` is the maximum number of iterations you're allowed overall in the case of failing to find a match, but you're only allowed to take a maximum of `call_limit.1` steps _after_ you've found the first match.

alexanderivrii replied:
Thanks. The code comment makes perfect sense now.

jakelishman replied:
Fixed in 8e8fdce51b.

alexanderivrii commented on crates/circuit/src/vf2.rs:
```diff
@@ -1233,10 +1233,10 @@ where
     node_semantics: NS,
     edge_semantics: ES,
-    restriction: Option<Restriction<NS::Score>>,
+    pub restriction: Option<Restriction<NS::Score>>,
```
why do we need to make `restriction` public here?

jakelishman replied:
I think maybe a mistake left over from an old commit.  Removed in c614cf45d3.

alexanderivrii commented on crates/transpiler/src/passes/vf2/vf2_layout.rs:
```diff
@@ -569,14 +580,22 @@ where
+        vf2.remaining_calls = config
+            .call_limit
+            .1
+            .map(|new| new.min(vf2.remaining_calls.unwrap_or(new)));
```
Shouldn't this simply be `vf2.remaining_calls = config.call_limit.1`?

jakelishman replied:
That could cause the limit to double in the case of `call_limit = (1000, 1000)` - you could use `999` trials, find a solution, then yours would cause you to gain another 1000 trials, but the intention is only to have 1.

alexanderivrii replied:
> That could cause the limit to double in the case of call_limit = (1000, 1000) - you could use 999 trials, find a solution, then yours would cause you to gain another 1000 trials, but the intention is only to have 1.

That makes sense, so the first number in the call_limit pair is indeed the maximum budget for the whole run. 

Hmm, but then setting `call_limit = (1000, None)` should also correspond to limiting the total budget to 1000, yet I believe the rust code would set `vf2.remaining_calls` to `None`.

jakelishman replied:
I rewrote the internals a little in 8e8fdce51b so that the bit we update is now called `vf2.call_limit`, and we separately track `vf2.num_calls`, which makes all this logic far more obvious and reliable.

alexanderivrii reviewed (commented):
It looks good (this is by far one of the easiest PRs in this chain 😄), but I did get confused about whether the first `call_limit` in the pair is supposed to represent the total budget for the VF2 algorithm (including the budget to find the first solution and to minimize it) or just the budget to find the first solution. Could you please clarify?

jakelishman committed:
Fix call-limit unbound conversion

jakelishman committed:
Remove unnecessary pub

jakelishman committed:
Revert separation of O1 and O2 VF2 configuration

That shouldn't be being done in this PR, if at all.

alexanderivrii reviewed (commented):
Just one more question about the computation of the remaining call limit in the case the second limit in the pair is `None`.

jakelishman committed:
Merge remote-tracking branch 'ibm/main' into vf2/prune/6

jakelishman committed:
Clarify `call_limit` meaning

Previously it was not clear that it's just two limits and we swap from
one to the other at the point of the first match.  The previous switch
was also not entirely correct, since swapping from a `None` to a `Some`
limit would not have counted calls already made.

alexanderivrii commented on qiskit/transpiler/preset_passmanagers/builtin_plugins.py:
```diff
@@ -747,9 +747,8 @@ def _swap_mapped(property_set):
                 coupling_map=pass_manager_config.coupling_map,
                 seed=-1,
-                call_limit=int(5e4),  # Set call limit to ~100ms with rustworkx 0.10.2
+                call_limit=(50_000, 1_000),
```
Just to make sure we are on the same page. Specifying `call_limit=(50_000, 1_000)` means:
* If we did not find the first layout within `50_000` calls, stop the transpiler pass.
* If finding the first layout required `0 < k < 1_000` calls, then we still have a budget of `1_000 - k` calls for minimizing the layout. 
* But if finding the first layout required `1_000 < k <= 50_000` calls, then we don't have budget for minimizing the layout.

I guess I am still wondering why, as a user, I would be willing to spend `50_000` calls for finding a solution, but I am willing to minimize the solution only if the first solution is found really quickly.

jakelishman replied:
The idea is that the first number is the upper bound on the runtime we'll spend in the worst case, but the second number is the runtime we'd _prefer_ to spend. Before this PR chain, that was done by a combination of `call_limit` and `max_trials`, but now `max_trials` is kind of meaningless, so this two-stage form is the replacement.

The new two-stage form is more reliable from a timing perspective in many ways, because the smaller limit is independent of how long it takes to find each trial.

alexanderivrii reviewed (commented):
Thanks for adding the clarification, now the meaning of both call limits is clear (and actually a little bit different from what I thought previously).

I have left one question inline, which essentially pokes at your intuition of why specifying a lower limit as the second limit is practically useful.

alexanderivrii reviewed (approved):
For sure, `call_limit` is much-much better than `max_trials`. The intuition "worst-case budget" vs. "preferred budget" does make sense to me, though in theory I could see other possible strategies (like: if one has the worst-case budget of 50_000 calls, prefers to spend 1_000 calls, and has already spent 20_000 calls for finding the first solution, then spend an additional 1_000 calls on improving that solution). I guess it boils down to how often and how quickly vf2 can improve the first solution on practical problems of interest.
